packages/common/src/util.js

```diff
@@ -49,3 +68,99 @@ export function normalizeOauthConfig(configuration) {
 
   return configuration;
 }
+
+function separateUrl(fullUrl) {
+  const urlObject = new URL(fullUrl);
+  const baseUrl = `${urlObject.protocol}//${urlObject.host}`;
+  const path = urlObject.pathname;
+  return { baseUrl, path };
+}
+
+export async function request(method, fullUrl, options = {}) {
+  const { baseUrl, path } = separateUrl(fullUrl);
+  const defaultOptions = {
+    headers: {},
+    query: undefined,
+    timeout: {},
+    tls: {},
+    body: undefined,
+    errors: {
+      401: 'Unauthorised',
+      404: 'Not found',
+    },
+  };
+
+  const { headers, query, body, errors, timeout, tls } = {
+    ...defaultOptions,
+    ...options,
+  };
+
+  if (!headers['Content-Type'] && body) {
+    headers['Content-Type'] = 'application/json';
+  }
+
+  const client = getClient(baseUrl);
+
+  const response = await client.request({
+    path: path,
+    query: query,
+    method: method,
+    headers: headers,
+    errors: errors,
+    body: body ? JSON.stringify(body) : undefined,
+    throwOnError: false,
+  });
+
+  const responseBody = await readResponseBody(response);
+
+  if (response.statusCode >= 400) {
+    const errorMessage = errors[response.statusCode]
+      ? errors[response.statusCode]
+      : `${method} request to ${path} failed with status: ${response.statusCode}`;
+
+    const error = new Error(
+      `Request to ${fullUrl} failed with status: ${response.statusCode}`
+    );
+    error.code = response.statusCode;
+    error.url = fullUrl;
+    error.message = errorMessage;
+    throw error;
+  }
+
+  return {
+    code: response.statusCode,
+    headers: response.headers,
+    data: responseBody,
+  };
+}
+
+async function readResponseBody(response) {
+  const contentType = response.headers['content-type'];
+
+  if (contentType?.includes('application/json')) {
+    return await response.body.json();
+  } else {
+    const chunks = [];
+    for await (const chunk of response.body) {
+      chunks.push(chunk);
+    }
+    return Buffer.concat(chunks).toString();
+  }
+}
+
+// Functions for different HTTP methods
```

Comment:
No need for this comment.

Also I think you can make the signatures more lightweight, like this:
```
export const get = (url, options) => request('GET', url, options)

export const post = (url, body, options) => request('GET', url, { body, ...options })
```

No need to default options here by the way, because the `request` will do that for us

Reply:
Thanks 👍🏽 , improve the function signature here [3ffec9a](https://github.com/OpenFn/adaptors/pull/339/commits/3ffec9a5a43d478ac4599488f4a9964a4aa43412)